### ISSUE TITLE
Allow tinting of a tile/object

### ISSUE THREAD

**Issue by Quillraven** [mapeditor/tiled#3950]
**Is your feature request related to a problem? Please describe.**
I am using Tiled also as an "object editor" for my game. My workflow is:
- Create a tileset with a collection of images
- Add an image to the tileset (e.g. the first frame of the idle animation of a game object like the player)
- Use collision editor to create Box2D fixture definitions
- Assign a class to the tile to add game properties to it, like damage, speed, ...

One of the properties is a color because it is possible to tint a specific game object. From what I know, this is currently not possible.
You can set a color for the class, which is not correct in my case because it is a general class for all game objects.
You can set a color on an object layer which is also not correct because I need a color for a specific object that is related to a specific tile of a tileset.

**Describe the solution you'd like**
Ideally, we can have a color attribut for a class with a specific key that will be used for tinting and which is then also considered for the tile rendering in Tiled (tile in tileset + object on objectLayer).
Maybe "tintColor" as a key or just "tint".

Since you can theoretically have multiple color attributes for a class, Tiled somehow needs to know which color it should use for tinting. That's why a special key might be useful or a separate property type "tint color" that can only be added once to a class.

**Describe alternatives you've considered**
I think there are no alternatives?

**Comment by eishiya** [comment]
Perhaps I misunderstood what you meant by "considered for the tile rendering in Tiled (tile in tileset + object on objectLayer)", but if you want all instances of a tile tinted, wouldn't it be better to tint the tile in its image instead?

It's worth noting that class colour is NOT a tint colour, so even if you assigned unique classes to each Tile Object, they'd not be tinted those colours. Class colours are mainly used to distinguish non-Tile Objects from each other. Similarly, the Color attribute on Object Layers does not tint Tile Objects, it just sets the default outline colour for other Object Types.

Only the Tint Color layer attribute actually changes the appearance of Tile Objects, and it is currently per-layer. So, currently, the only way to customise your Tile Object tints in Tiled is to put the Tile Objects on layers corresponding to their tints, you'd need as many Object Layers as you have different tint colours.

I support the idea of a **per-Object Colour** (that is, a new attribute on Objects, not on Tiles), functioning like so:
- On Tile Objects, tint the artwork (multiply by that colour).
- On other Objects, override the colour from their class or Object Layer. Perhaps there could be a View or Preferences option to multiply the two colours instead, but in the scenarios people have requested per-Object colour before, I think overriding would be more useful.
- Could be stored as just a `color` attribute on the Object.

**Comment by Quillraven** [comment]
Imo adjusting the color of the artwork is not ideal. Imagine you have an art for a dragon in white color. Now, you want to use the same art three different times. Once for the original white dragon, a red version for a fire dragon and a blue version for a water dragon. Like "reskinning" your existing assets.

Most likely you'd solve that by simply tinting the graphic for the different dragons as this saves size on your texture/texture atlas in the game.
Also, it is way quicker to tint the graphic instead of opening your editor -> adjust the color -> save the file.

That's why imo the coloring per tile would be cool. Maybe tile is the wrong wording here. As mentioned, I am using a "collection of image" tileset in my workflow.

In the scenario above I would import the dragon image three times into this tileset. Two images will have a tint color for red and blue.
Ideally, I already see the tinted version in the tileset. Also, ideally I see the same representation in a map itself that uses the image of that tileset.

If it is still not clear, I will try to make some screenshots tomorrow but hopefully I explained it well enough :)

**Comment by eishiya** [comment]
Sorry, I neglected the possibility that you might have the same base image added to a tileset multiple times - I usually work with "Based on Tileset Image" tilesets, so I'm used to it being impossible for two different tiles sharing the same source pixels.

**Comment by Quillraven** [comment]
Hi,

Is there an progress on this? Or what is the priority of open issues?

It is nothing critical or urgent but just interested in how long it usually takes to get something like this available.

If I can help with something just let me know!

Thank you 😊

**Comment by bjorn** [comment]
> Is there an progress on this? Or what is the priority of open issues?

There's currently no priority on this particular issue and no progress has been made. I think you're also the only one to request per-tile tinting, but I could be wrong. Per-object tinting has been requested (though I couldn't find an issue about it now) and I guess is even halfway implemented since it's basically an extension of the unfinished https://github.com/mapeditor/tiled/pull/4031 that adds per-object opacity.

That said, I think per-tile tinting would be a nice feature to add and does not conflict with per-object tinting. The eventual object tint could likely just multiply with the tile tint, like it would with the layer tint.

I do try to listen to what users are requesting, but apart from just giving feedback there's two concrete ways to prioritize an issue, which is to work on implementing it yourself or by donating on a tier that grants development time towards your request. I'm available (also on Discord) in case you have any questions about how to implement this.

**Comment by Quillraven** [comment]
Thanks for the quick response! Unfortunately, I think Tiled will take quite a lot of time for me to know how to implement something and I won't have the time to help.

But again, it is nothing urgent or critical. I was just curious how things are prioritized.

And indeed I am a donator but not a huge one 😅 But I really appreciate the editor and all it can do. It is a great tool 👍 thanks for all the effort.

**Comment by LJNIC** [comment]
I would also like this feature. Can't really find a "game oriented" editor that supports this. REXPaint, Playscii, lvllvl all have this but none of the metadata features.

**Comment by hiive** [comment]
This would be really useful on tiles and objects for all of the reasons described above.

**Comment by YusefZahran** [comment]
I found this issue/feature and have been thinking about ways to implement it but almost all solutions i thought of have their pros and cons. I first looked into whether there's a class I could implement tinting so it affects both tiles/cells and Object, like `Tile::Object`, but that might not be a great idea since a lot of classes inherit from them, so I scrapped that and looked for other ideas.

The next idea would to implement tinting for `MapObjects` and Tiles separately. For `MapObjects` this would be relatively easy, but for tiles, not so much.

From my understanding, we're trying to tint individual cells placed on the map so they could have their separate tints if the user desires. We could potentially do this by adding a color field/variable to the `Cell` class, but with a map having from hundreds to thousands or even potentially millions of cells, it would quickly start taking a lot of memory.
(I haven't calculated how much of a memory increase it would actually be and whether it would cause problems or not, but I'm also not sure that we want to complicate the relatively simply Cell struct, and potentially having to rework a lot of core logic and introduce compatibility issues)

The Second idea was to build a data structure that stores which cells the user tinted, and check it each render to change the cell's color. This would solve the memory problem form the previous idea, but would be a performance hit, having to check the data structure each draw call.

The third idea I came up with, is each time a user decides to tint a cell, it would clone the tile from the tileset, tint/color it, and add it to the tileset as a tinted copy. This would be like adding a new tile to the tileset, just it being unique from the original tile. There could also be a separate section in the tileset view for these tinted/modified tiles. 
I haven't researched too deep into this idea and how feasible it is to implement (perhaps there could be problem differentiating the original tile and new modified ones, and i don't know if all tiles must have an origin from their reference image source, as some game engines do look up from them, so there might be some export probelms to find solutions to), but if it is indeed feasible it would a pretty good and user friendly solution.

@bjorn I'd like to hear your thoughts and feedback on this, as i think color/tinting would be a pretty nice feature for tiled to have, but it should be approached and implemented carefully.

**Comment by bjorn** [comment]
@YusefZahran Yes, I think tinting per `Cell` is going too far. Indeed it would grow memory usage even though it won't be used by most people, and it would also require additional tooling to "paint" with colors on the map. It does not sound like this issue even asked for something like that.

What this issue is about is to considering supporting a tint color on:

* Tiles in a tileset (it is possible to duplicate a tile when it needs to be used in several colors, as mentioned above - soon also for "atlas tilesets" as implemented in #4166).
* Map objects (tile objects) placed on an object layer.

In both cases, the "tint color" property can simply be added to the Properties view, like it already is for Layer Properties since we already support a tint color on layers. We do need to give some thought about how these colors might be combined (if all these properties are implemented, a tile object might be tinted by its tile, its layer and itself. They could either multiply or override each other. I think the layer tint should probably multiply, but I'm not sure the same is useful for map object tint vs. tile tint, where we also have precedence for other overriding properties (class and custom properties).

**Comment by Quillraven** [comment]
Hi, since that issue is from me I just wanted to point out that Bjorn got it right. No need for individual cells to get tinted. Very brief why I raised that issue:
- I was working on an RPG and used collection of images tileset to configure enemies (properties + collision objects)
- I added some images twice (a stronger enemy of the same type)
- in game I tinted that in different colors to be able to distinguish which enemy version it is
- in Tiled it was just a color property which unfortunately has no effect on the tile in the tileset itself
- those tiles are actually not placed on the map. They are just "config objects". Random encounters then spawn enemies based on these configs

So I personally was missing a functionality to tint a tile of a collection of images tileset. That was the origin of this issue.

Hope that helps!

**Comment by YusefZahran** [comment]
Hello! This took me a while (working on a feature across like 20+ files in an unfamiliar codebase is hard). I've implemented tinting for both tileObjects and tiles, and I'm to ready to push a PR, but I'm unsure if it's better to have them both in one PR since they're very similar in logic, or separate them (tileObjects tinting, tile tinting), as they're technically different features. @bjorn Some guidance would be appreciated.
